File: QytetetRuby/pkg/QytetetRuby-0.0.1/lib/qytetet.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require "singleton"
require_relative "dado"
require_relative "tablero"
require_relative "sorpresa"
require_relative "casilla"
require_relative "jugador"
require_relative "titulo_propiedad"

module ModeloQytetet

  class Qytetet
    include Singleton
    
    attr_reader :cartaActual, :jugadorActual
    
    @@MAX_JUGADORES = 4
    @@MAX_CARTAS = 10
    @@MAX_CASILLAS = 20
    @@PRECIO_LIBERTAD = 200
    @@SALDO_SALIDA = 1000
    
    def initialize
      @cartaActual = nil
      @mazo = []
      @jugadores = []
      @jugadorActual = nil
      @tableroActual = nil
      @dado = Dado.instance
      
    end
    
    def inicializar_cartas_sorpresa()
      @mazo << Sorpresa.new("Te hemos pillado con chanclas y calcetines. Lo sentimos, ¡debes ir a la carcel!", @tableroActual.carcel.numeroCasilla, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("¡Un OVNI te ha abducido y te ha llevado hasta la casilla de SALIDA! Cobra tu recompensa por sobrevivir.",0, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Un fan anónimo ha pagado tu fianza. Sales de la cárcel", 0, TipoSorpresa::SALIRCARCEL)
      @mazo << Sorpresa.new("Gracias a tus poderosos contactos quedas libre de la cárcel", 0, TipoSorpresa::SALIRCARCEL)
      @mazo << Sorpresa.new("¡Levantas una piedra y encuentras dinero!", 200, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Por no recoger los excrementos de tu perro pagas a cada jugador por las molestias", 100, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Eres un químico muy patoso y provocas una explosión que afecta a todo el vecindario. Pagas por cada casa/hotel que haya en tu misma arista del tablero", 100, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("Recibes el pago mensual por alquiler de casas y reservas de hotel. Recibes:", 150, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("Haces una fiesta privada y asisten todos los jugadores. El precio de la entrada es:", 150, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("¡Se te va de las manos! Por escándalo público pagas: ", 150, TipoSorpresa::PORJUGADOR)
      
      @mazo.shuffle
    end
    
    def inicializar_jugadores nombres
      for name in nombres
        @jugadores << Jugador.new(name)
      end
#      if (i<1 || i>MAX_JUGADORES-1)
#        raise UnsupportedOperationException, 'Sin implementar'
#      end
    end
    
    def inicializar_tablero
      @tableroActual = Tablero.new
    end
    
    def inicializar_juego nombres
      inicializar_tablero()
      inicializar_jugadores (nombres)
      inicializar_cartas_sorpresa()
      salida_jugadores()
    end
    
    def get_carta_actual
      return @cartaActual
    end
    
    def get_jugador_actual
      return @jugadorActual
    end
    
    def comprar_titulo_propiedad
      @jugadorActual.comprar_titulo()
    end
    
    def edificar_casa casilla
      puedo_edificar = false
      
      if(casilla.soy_edificable())
        se_puede_edificar = casilla.se_puede_edificar_casa()
        
        if(se_puede_edificar)
          puedo_edificar = @jugadorActual.puedo_edificar_casa(casilla)
          
          if(puedo_edificar)
            coste_edificar_casa = casilla.edificar_casa()
            @jugadorActual.modificar_saldo(-coste_edificar_casa)
          end
        end
      end
      
      puedo_edificar
    end
    
    def edificar_hotel casilla
      puedo_edificar = false
      
      if(casilla.soy_edificable())
        se_puede_edificar = casilla.se_puede_edificar_hotel()
        
        if(se_puede_edificar)
          puedo_edificar = @jugadorActual.puedo_edificar_hotel(casilla)
          
          if(puedo_edificar)
            coste_edificar_hotel = casilla.edificar_hotel()
            @jugadorActual.modificar_saldo(-coste_edificar_hotel)
          end
        end
      end
      
      puedo_edificar
    end
    
    def hipotecar_propiedad casilla
      puedo_hipotecar_propiedad = false
      
      if(casilla.soy_edificable())
        if(!casilla.esta_hipotecada())
          if(@jugadorActual.puedo_hipotecar(casilla))
            cantidad_recibida = casilla.hipotecar()
            @jugadorActual.modificar_saldo(cantidad_recibida)
            puedo_hipotecar_propiedad = true
          end
        end
      end
      
      puedo_hipotecar_propiedad
    end
    
    def cancelar_hipoteca casilla
      puedo_cancelar = @jugadorActual.puedo_pagar_hipoteca(casilla)
      if(puedo_cancelar)

          coste_hipoteca = casilla.cancelar_hipoteca()
          @jugadorActual.modificar_saldo(-coste_hipoteca)
      end

      puedo_cancelar
    end
    
    def vender_propiedad casilla
      puedo_vender = false
      if (casilla.soy_edificable())
        puedo_vender = @jugadorActual.puedo_vender_propiedad(casilla)
        
        if(puedo_vender)
          @jugadorActual.vender_propiedad(casilla)
        end
      end
      puedo_vender
    end
    
    def encarcelar_jugador
      if(!@jugadorActual.tengo_carta_libertad())
        
        casilla_carcel = @tableroActual.carcel
        
        @jugadorActual.ir_a_carcel(casilla_carcel)
        
      else
        
        carta = @jugadorActual.devolver_carta_libertad()
        mazo << carta
        
      end
    end
    
    def aplicar_sorpresa
      tienePropietario = false
      
      if (@cartaActual.tipo == TipoSorpresa::PAGARCOBRAR)
        
        @jugadorActual.modificar_saldo(@cartaActual.valor)
        
      elsif (@cartaActual.tipo == TipoSorpresa::IRACASILLA)
        
        es_carcel = @tableroActual.es_casilla_carcel(@cartaActual.valor)
        
        if(es_carcel)
          
          encarcelar_jugador()
          
        else
          
          nueva_casilla = @tableroActual.obtener_casilla_numero(@cartaActual.valor)
          tiene_propietario = @jugadorActual.actualizar_posicion(nueva_casilla)
          
        end
        
      elsif (@cartaActual.tipo == TipoSorpresa::PORCASAHOTEL)
        
        @jugadorActual.pagar_cobrar_por_casa_hotel(@cartaActual.valor)
        
      elsif(@cartaActual.tipo == TipoSorpresa::PORJUGADOR)
        
        for jug in @jugadores
          
          if (jug != @jugadorActual)
            
            jug.modificar_saldo(@cartaActual.valor)
            @jugadorActual.modificar_saldo(-@cartaActual.valor)
            
          end
          
        end
        
      end
      
      if(@cartaActual.tipo == TipoSorpresa::SALIRCARCEL)
        
        @jugadorActual.cartaLibertad = @cartaActual
      end  
        
      @mazo << @cartaActual  #SOLO SE DEVUELVE LA CARTA SI ES TIPO CACEL??
        
      
    end
    
    def intentar_salir_carcel metodo
      libre = false
      
      if(metodo == MetodoSalirCarcel::TIRANDODADO)
        valor_dado = @dado.tirar()
        libre = valor_dado > 5
        
      elsif (metodo == MetodoSalirCarcel::PAGANDOLIBERTAD)
        libre = @jugadorActual.pagar_libertad(-@@PRECIO_LIBERTAD)
      end
      
      if (libre)
        @jugadorActual.encarcelado = false
      end
      libre
    end
    
    def jugar
      tiene_propietario = false
      
      puts "\n\n\n****Pulsa Intro para tirar el dado****\n\n"
      gets.chomp
      valor_dado = @dado.tirar()
      puts "Has sacado #{valor_dado}\n\n\n"
      casilla_posicion = @jugadorActual.casillaActual
      nueva_casilla = @tableroActual.obtener_nueva_casilla(casilla_posicion, valor_dado)
      
      tiene_propietario = @jugadorActual.actualizar_posicion(nueva_casilla)
      
      if(!nueva_casilla.soy_edificable())
        
        if(nueva_casilla.tipo == TipoCasilla::JUEZ)
          
          encarcelar_jugador()
          
        elsif(nueva_casilla.tipo == TipoCasilla::SORPRESA)
          
          @cartaActual = @mazo[0]
          puts @cartaActual.texto
          @mazo.delete_at(0)
          
        end
        
      end
      
      
      tiene_propietario
    end
    
    def obtener_ranking
      ranking = []
      aux = @jugadores
      
      while(aux.size > 0)
        
        max = aux[0]
        
        for jug in aux
          
          if(jug.obtener_capital() > max.obtener_capital())
            
            max = jug
            
          end
          
          ranking << max
          aux.delete(max)
          
        end
        
      end
      
      ranking
      
    end
    
    def siguiente_jugador
      i = 0
      seguir = true
      
      while (i < @jugadores.size && seguir)
        
        if(@jugadores[i] == @jugadorActual)
          seguir = false
          @jugadorActual = @jugadores[ ( (i+1) % @jugadores.size ) ]
        end
        i = i+1
      end
    
    end
    
    def salida_jugadores
      for jug in @jugadores
        jug.saldo = 7500
        jug.casillaActual = @tableroActual.obtener_casilla_numero(0)
      end
      
      @jugadorActual = @jugadores[rand(@jugadores.size)]
    end
    
    def propiedades_hipotecadas_jugador hipotecada
=begin      devolver = []
      
      for jug in @jugadores
        if ( jug.propiedades.hipotecada == hipotecada )
          devolver << jug.propiedades.casilla
        end
      end
      
      devolver
=end
      @jugadorActual.propiedades_hipotecadas(hipotecada)
    end
    
    def get_jugadores
      @jugadores
    end
    
    def to_s
      resultado = "Qytetet!! \n\n Jugadores: "
      for j in @jugadores
        resultado += j.to_s
      end
      resultado += "\n\n Mazo: \n"
      for m in @mazo
        resultado += m.to_s
      end
      resultado += "\n\n      Tablero:\n         #{@tableroActual} \n JugadorActual: #{@jugadorActual} \n CartaSorpresaActual: #{@cartaActual}\n"
    end
    
    public :cartaActual, :jugadorActual
  end
  
end

File: PRACTICAS/QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import GUIQytetet.Dado;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

/**
 *
 * @author jjavier98
 */
public class Qytetet {

    private static Qytetet instance = new Qytetet();

    public static Qytetet getInstance() {
        return instance;
    }

    public static final int MAX_JUGADORES = 4;
    public static final int MAX_CARTAS = 10;
    public static final int MAX_CASILLAS = 20;
    public static final int PRECIO_LIBERTAD = 200;
    public static final int SALDO_SALIDA = 7500;

    private Sorpresa cartaActual;
    private final ArrayList<Sorpresa> mazo;
    private ArrayList<Jugador> jugadores;
    private Jugador jugadorActual;
    private Tablero tablero;
    private Dado dado;

    private Qytetet() {
        cartaActual = null;
        mazo = new ArrayList<>();
        jugadores = new ArrayList<>();
        jugadorActual = null;
        dado = GUIQytetet.Dado.getInstance();
        tablero = null;
    }

    private void inicializarCartasSorpresas() {
        mazo.add(new Sorpresa("Te has convertido en un jugador ESPECULADOR tipo 1!", TipoSorpresa.CONVERTIRME, 3000));
        mazo.add(new Sorpresa("Te has convertido en un jugador ESPECULADOR tipo 2!!", TipoSorpresa.CONVERTIRME, 5000));
        mazo.add(new Sorpresa("Gracias a tus poderosos contactos quedas libre de la cárcel", TipoSorpresa.SALIRCARCEL, 0));
        mazo.add(new Sorpresa("Un fan anónimo ha pagado tu fianza. Sales de la cárcel", TipoSorpresa.SALIRCARCEL, 0));
        mazo.add(new Sorpresa("Te hemos pillado con chanclas y calcetines. Lo sentimos, ¡debes ir a la carcel!", TipoSorpresa.IRACASILLA, tablero.getCarcel().getNumeroCasilla()));
        mazo.add(new Sorpresa("¡Un OVNI te ha abducido y te ha llevado hasta la casilla de SALIDA! Cobra tu recompensa por sobrevivir.", TipoSorpresa.IRACASILLA, 0));
        mazo.add(new Sorpresa("Gracias a tus poderosos contactos quedas libre de la cárcel", TipoSorpresa.SALIRCARCEL, 0));
        mazo.add(new Sorpresa("¡Levantas una piedra y encuentras dinero!", TipoSorpresa.PAGARCOBRAR, 300));
        mazo.add(new Sorpresa("Por no recoger los excrementos de tu perro pagas a cada jugador por las molestias", TipoSorpresa.PAGARCOBRAR, -100));
        mazo.add(new Sorpresa("Eres un químico muy patoso y provocas una explosión que afecta a todo el vecindario. Pagas por cada casa/hotel que haya en tu misma arista del tablero", TipoSorpresa.PORCASAHOTEL, 100));
        mazo.add(new Sorpresa("Recibes el pago mensual por alquiler de casas y reservas de hotel.", TipoSorpresa.PORCASAHOTEL, -150));
        mazo.add(new Sorpresa("Haces una fiesta privada y asisten todos los jugadores. El precio de la entrada es:", TipoSorpresa.PORJUGADOR, -150));
        mazo.add(new Sorpresa("¡Se te va de las manos! Por escándalo público pagas: ", TipoSorpresa.PORJUGADOR, 150));

        //Collections.shuffle(mazo);
    }

    private void inicializarJugadores(ArrayList<String> nombres) {
        for (String nom : nombres) {
            jugadores.add(new Jugador(nom));
        }
    }

    private void inicializarTablero() {
        tablero = new Tablero();
    }

    public boolean comprarTituloPropiedad() {
        return jugadorActual.comprarTitulo();
    }

    public boolean hipotecarPropiedad(Calle casilla) {
        boolean puedoHipotecarPropiedad = false;

        if (casilla.soyEdificable()) {
            boolean sePuedeHipotecar = !casilla.estaHipotecada();

            if (sePuedeHipotecar) {
                if (jugadorActual.puedoHipotecar(casilla)) {
                    puedoHipotecarPropiedad = true;
                    int cantidadRecibida = casilla.hipotecar();
                    jugadorActual.modificarSaldo(cantidadRecibida);
                }
            }
        }

        return puedoHipotecarPropiedad;

    }

    public boolean cancelarHipoteca(Calle casilla) {
        boolean puedoCancelar = jugadorActual.puedoPagarHipoteca(casilla);
        if (puedoCancelar) {
            int costeHipoteca = casilla.cancelarHipoteca();
            jugadorActual.modificarSaldo(-costeHipoteca);
        }

        return puedoCancelar;
    }

    public void inicializarJuego(ArrayList<String> nombres) {
        inicializarTablero();
        inicializarJugadores(nombres);
        inicializarCartasSorpresas();
        salidaJugadores();
    }

    public Sorpresa getCartaActual() {
        return cartaActual;
    }

    public boolean edificarCasa(Calle casilla) {
        boolean puedoEdificar = false;

        if (casilla.soyEdificable()) {
            boolean sePuedeEdificar = casilla.sePuedeEdificarCasa();
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificarCasa(casilla);

                if (puedoEdificar) {
                    int costeEdificarCasa = casilla.edificarCasa();
                    jugadorActual.modificarSaldo(-costeEdificarCasa);
                }
            }
        }
        return puedoEdificar;
    }

    public boolean edificarHotel(Calle casilla) {
        boolean puedoEdificar = false;

        if (casilla.soyEdificable()) {
            boolean sePuedeEdificar = casilla.sePuedeEdificarHotel();
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificarHotel(casilla);

                if (puedoEdificar) {
                    int costeEdificar = casilla.edificarHotel();
                    jugadorActual.modificarSaldo(-costeEdificar);
                }
            }
        }
        return puedoEdificar;
    }

    public Jugador getJugadorActual() {
        return jugadorActual;
    }

    public void siguienteJugador() {
        boolean seguir = true;

        for (int i = 0; i < jugadores.size() && seguir; ++i) {
            if (jugadores.get(i) == jugadorActual) {
                seguir = false;
                jugadorActual = jugadores.get((i + 1) % jugadores.size());
            }
        }
    }

    public boolean venderPropiedad(Calle casilla) {
        boolean puedoVender = false;

        if (casilla.soyEdificable()) {
            puedoVender = jugadorActual.puedoVenderPropiedad(casilla);

            if (puedoVender) {
                jugadorActual.venderPropiedad(casilla);
            }
        }

        return puedoVender;
    }

    public boolean aplicarSorpresa() {
        boolean tienePropietario = false;

        TipoSorpresa tipo = cartaActual.getTipo();

        System.out.println(cartaActual.getTexto());

        if (null != tipo) {
            switch (tipo) {
                case PAGARCOBRAR:
                    jugadorActual.modificarSaldo(cartaActual.getValor());
                    break;
                case IRACASILLA:
                    boolean esCarcel = tablero.esCasillaCarcel(cartaActual.getValor());
                    if (esCarcel) {
                        encarcelarJugador();
                    } else {
                        Casilla nuevaCasilla = tablero.obtenerCasillaNumero(cartaActual.getValor());
                        tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
                    }
                    break;
                case PORCASAHOTEL:
                    jugadorActual.pagarCobrarPorCasaYHotel(cartaActual.getValor());
                    break;
                case PORJUGADOR:
                    for (Jugador jug : jugadores) {
                        if (jugadorActual != jug) {
                            jug.modificarSaldo(cartaActual.getValor());
                            jugadorActual.modificarSaldo(-cartaActual.getValor());
                        }
                    }
                    break;
                case CONVERTIRME:
                    Especulador espec = jugadorActual.convertirme(cartaActual.getValor());
                    boolean encontrado = false;
                    for (int i = 0; i < jugadores.size() && !encontrado; i++) {
                        if (jugadores.get(i) == jugadorActual) {
                            encontrado = true;
                            jugadores.set(i, espec);
                            jugadorActual = jugadores.get(i);
                        }
                    }
                default:
                    break;
            }
        }

        if (tipo == TipoSorpresa.SALIRCARCEL) {
            jugadorActual.setCartaLibertad(cartaActual);
        } else {
            mazo.add(cartaActual); //SOLO SE DEVUELVE LA CARTA SI ES TIPO CACEL??
        }

        return tienePropietario;
    }

    private void salidaJugadores() {
        for (Jugador jug : jugadores) {
            jug.setCasillaActual(tablero.obtenerCasillaNumero(0));
            jug.setSaldo(SALDO_SALIDA);
        }
        int numero = (int) (Math.random() * jugadores.size());
        jugadorActual = jugadores.get(numero);
    }

    public boolean jugar() {
        Scanner scan = new Scanner(System.in);

        boolean tienePropietario;

        Dado dado = GUIQytetet.Dado.getInstance();
        int valorDado = dado.nextNumber("DADO QYTETET", "¡Tira el dado!");
        System.out.println("Has sacado: " + valorDado);

        Casilla casillaPosicion = jugadorActual.getCasillaActual();
        Casilla nuevaCasilla = tablero.obtenerNuevaCasilla(casillaPosicion, valorDado);
        
        if(nuevaCasilla.getTipo() != TipoCasilla.CALLE)
            System.out.println("\nCaes en la casilla:\n   *númeroCasilla: " +
                    nuevaCasilla.getNumeroCasilla() + "\n   *Tipo: " + nuevaCasilla.getTipo() + "\n");
        else
            if(nuevaCasilla.getTipo() == TipoCasilla.CALLE && nuevaCasilla.tengoPropietario())
                System.out.println("\nCaes en la casilla:\n   *númeroCasilla: " +
                        nuevaCasilla.getNumeroCasilla() + "\n   *Tipo: " + nuevaCasilla.getTipo() +
                        "\n   *Propietario: " + ((Calle)nuevaCasilla).getTitulo().getPropietario().getNombre() + "\n");
        
        tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);

        if (!nuevaCasilla.soyEdificable()) {
            if (nuevaCasilla.getTipo() == TipoCasilla.JUEZ) {
                encarcelarJugador();
            } else if (nuevaCasilla.getTipo() == TipoCasilla.SORPRESA) {
                cartaActual = mazo.get(0);
                mazo.remove(0); // SE QUITA PORQUE AL DEVOLVER LA CARTA SE AÑADIRÁ AL FINAL DEL MAZO
            }
        }

        return tienePropietario;
    }

    public ArrayList<Jugador> obtenerRanking() {
        ArrayList<Jugador> ranking = new ArrayList<>();
        ArrayList<Jugador> aux = new ArrayList<>();
        aux = jugadores;

        while (aux.size() > 0) {
            Jugador max = aux.get(0);

            for (Jugador jug : aux) {
                if (jug.obtenerCapital() > max.obtenerCapital()) {
                    max = jug;
                }
            }
            ranking.add(max);
            aux.remove(max);
        }

        return ranking;
    }

    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo) {
        Scanner scan = new Scanner(System.in);
        boolean libre = false;

        if (metodo == MetodoSalirCarcel.TIRANDODADO) {
            int valorDado = dado.nextNumber("DADO QYTETET", "¡Tira el dado!");
            System.out.println("Has sacado: " + valorDado);
            libre = valorDado > 5;
        } else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD) {
            boolean tengoSaldo = jugadorActual.pagarLibertad(-PRECIO_LIBERTAD);
            libre = tengoSaldo;
        }

        if (libre) {
            jugadorActual.setEncarcelado(false);
        }

        return libre;
    }

    public void encarcelarJugador() {
        if (!jugadorActual.tengoCartaLibertad()) {
            Casilla casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
        } else {
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
        }
    }

    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas) {
        ArrayList<Casilla> devolver = new ArrayList<>();
        for (TituloPropiedad tp : jugadorActual.getPropiedades()) {
            if (tp.getHipotecada() == hipotecadas) {
                devolver.add(tp.getCasilla());
            }
        }
        return devolver;
    }

    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }

    @Override
    public String toString() {
        String resultado = "Qytetet!! \n\n Jugadores: ";
        for (Jugador j : jugadores) {
            resultado += j.toString();
            resultado += "\n";
        }
        resultado += "\n\n Mazo Sorpresas: \n";
        for (Sorpresa m : mazo) {
            resultado += m.toString();
        }
        resultado += "\n\n      Tablero:\n         " + tablero + "\n JugadorActual:" + jugadorActual + "\n CartaSorpresaActual:" + cartaActual + '\n';
        return resultado;
    }

}
